File: 64-Mnimum-Path-Sum/solution.cpp
class Solution {
    public:
        int minPathSum(vector<vector<int>>& grid) {

            int m = grid.size();
            int n = grid[0].size();

            vector<vector<int>> x(m, vector<int>(n, grid[0][0]));

            for (int i = 1; i < m; i++)
                x[i][0] = grid[i][0] + x[i-1][0];
            for (int i = 1; i < n; i++)
                x[0][i] = grid[0][i] + x[0][i-1];

            for (int i = 1; i < m; i++)
                for (int j = 1; j < n; j++)
                    x[i][j] = min(grid[i][j]+x[i-1][j], grid[i][j]+x[i][j-1]);

            return x[m-1][n-1];
        }

};

File: 64-Mnimum-Path-Sum/solution.py
import sys
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])

        dp = [[ 0 for i in range(n+1)] for j in range(m+1)]

        for i in range(len(dp)):
          dp[i][0] = sys.maxsize

        for i in range(len(dp[0])):
            dp[0][i] = sys.maxsize

        dp[1][1] = grid[0][0]

        for i in range(1, m+1):
            for j in range(1, n+1):
                if i == 1 and j == 1: 
                    continue
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i-1][j-1]


        return dp[-1][-1] if m and n else 0


grid = [[1,3,1],[1,5,1],[4,2,1]]

print(Solution().minPathSum(grid))
